### PR TITLE
Dev Day spec updates, including ingested /realtime definitions

### PR DESCRIPTION
**Pull request by trrwilson**

### CODE REVIEW

File: .dotnet/src/Custom/Moderations/ModerationInputKinds.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+using System;
+
+namespace OpenAI.Moderations;
+
+[Flags]
+public enum ModerationInputKinds : int
+{
+    Other = 1 << 0,
```
Comment (joseharriaga) [inline .dotnet/src/Custom/Moderations/ModerationInputKinds.cs:8]: We need a `None` value with value equal to 0.

File: .dotnet/src/Custom/Chat/ChatCompletionOptions.cs [hunk 2]
```diff
@@ -159,4 +159,10 @@ public ChatCompletionOptions()
     // CUSTOM: Added the Obsolete attribute.
     [Obsolete($"This property is obsolete. Please use {nameof(Tools)} instead.")]
     public IList<ChatFunction> Functions { get; }
+
+    [CodeGenMember("Metadata")]
+    public IDictionary<string, string> Metadata { get; } = new ChangeTrackingDictionary<string, string>();
```
Comment (joseharriaga) [inline .dotnet/src/Custom/Chat/ChatCompletionOptions.cs:164]: I can see that the "store" property needs to be renamed, but why does "metadata" need to be customized? 🤔 Could you add `// CUSTOM` comments for both properties?

File: .dotnet/src/Custom/Chat/ChatCompletionOptions.cs [hunk 3]
```diff
@@ -159,4 +159,10 @@ public ChatCompletionOptions()
...
+
+    [CodeGenMember("Store")]
+    public bool? EnableOutputStorage { get; set; }
```
Comment (joseharriaga) [inline .dotnet/src/Custom/Chat/ChatCompletionOptions.cs:167]: nit: Is calling it `StoreOutput` better? Basically following the pattern that we used for `IncludeLogProbabilities` and `AllowParallelToolCalls` in this class.

File: .dotnet/src/Custom/Chat/ChatFunctionCall.cs [hunk 4]
```diff
@@ -40,4 +40,24 @@ public partial class ChatFunctionCall
     /// </summary>
     [CodeGenMember("Arguments")]
     public BinaryData FunctionArguments { get; }
+
+    // CUSTOM: parameter order preserved.
+    public ChatFunctionCall(string functionName, BinaryData functionArguments)
```
Comment (joseharriaga) [inline .dotnet/src/Custom/Chat/ChatFunctionCall.cs:45]: Should we add a small customization in the TypeSpec to change the parameter order there so that we don't lose support for generated code?

File: .dotnet/src/Custom/Chat/ChatMessageContent.cs [hunk 5]
```diff
@@ -8,7 +9,7 @@ namespace OpenAI.Chat;
 public partial class ChatMessageContent : Collection<ChatMessageContentPart>
 {
     public ChatMessageContent()
-        : this([])
+        : this(Array.Empty<ChatMessageContentPart>())
```
Comment (joseharriaga) [inline .dotnet/src/Custom/Chat/ChatMessageContent.cs:12]: Was the original a bug? Are these not interchangeable? 🤔

File: .dotnet/src/Custom/Audio/AudioTranscriptionFormat.cs [hunk 6]
```diff
@@ -1,41 +1,48 @@
...
+// CUSTOM: Introduced as scenario-specific redirection of common format.
 /// <summary> The format of the transcription. </summary>
-[CodeGenModel("CreateTranscriptionRequestResponseFormat1")]
-public readonly partial struct AudioTranscriptionFormat
+public readonly partial struct AudioTranscriptionFormat : IEquatable<AudioTranscriptionFormat>
 {
-    // CUSTOM:
-    // - Applied the EditorBrowsable attribute.
-    // - Added custom doc comments.
+    internal readonly InternalAudioResponseFormat _internalAudioResponseFormat;
```
[the remark this answers is not in the document]
Comment (trrwilson) [reply .dotnet/src/Custom/Audio/AudioTranscriptionFormat.cs:10]: I can give that a shot. My concern is that we're growing our spec pollution debt (convoluting .tsp for emitter purposes) and there's increasing urgency/pressure/value for us to use this spec and an upgraded Azure counterpart to generate the Swagger documents and replace AOAI's current handwritten ones. This'll be a particularly egregious case since we'll end up with *three* copies of the same thing in the .yaml.

Still probably the right thing to do right now, though, and likely addressable via something holistic like using a different `client.tsp` for .cs emission and keeping all the pollution contained in that.

File: .dotnet/api/OpenAI.netstandard2.0.cs [hunk 7]
```diff
@@ -40,11 +40,6 @@ public class Assistant : IJsonModel<Assistant>, IPersistableModel<Assistant> {
         public float? Temperature { get; }
         public ToolResources ToolResources { get; }
         public IReadOnlyList<ToolDefinition> Tools { get; }
-        Assistant IJsonModel<Assistant>.Create(ref Utf8JsonReader reader, ModelReaderWriterOptions options);
```
Comment (joseharriaga) [inline .dotnet/api/OpenAI.netstandard2.0.cs:43]: Did you make a change to Export-API.ps1 to delete these?

File: .dotnet/src/Custom/Audio/AudioTranscriptionFormat.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (trrwilson) [reply .dotnet/src/Custom/Audio/AudioTranscriptionFormat.cs]: Updated -- we can eliminate the extra tool behavior once we have clarity on .tsp enum use and/or union spread support.

File: .dotnet/src/Custom/Moderations/ModerationCategory.cs [hunk 8]
```diff
@@ -1,13 +1,17 @@
...
+    internal ModerationCategory(bool flagged, float score, IEnumerable<string> internalAppliedInputTypes)
     {
         Flagged = flagged;
         Score = score;
+        InputKinds = ModerationInputKindsExtensions.FromInternalInputKinds(internalAppliedInputTypes);
     }
 
     public bool Flagged { get; }
     public float Score { get; }
+    public ModerationInputKinds InputKinds { get; }
```
[the remark this answers is not in the document]
Comment (trrwilson) [reply .dotnet/src/Custom/Moderations/ModerationCategory.cs:16]: Per off-PR discussion, what about something like `ModerationFlaggedInputKinds FlaggedInputKinds`? If we want to be bold and synchronize nomenclature we could also consider `ModerationFlaggedContentModalities FlaggedContentModalities`.

[the remark this answers is not in the document]
Comment (trrwilson) [reply .dotnet/src/Custom/Moderations/ModerationCategory.cs:16]: Updating to try out `ModerationFlaggedContentModalities FlaggedContentModalities` for now -- should at least be an improvement!

File: .dotnet/CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,23 @@
 # Release History
 
+## 2.1.0-beta.2 (Unreleased)
```
[the remark this answers is not in the document]
Comment (trrwilson) [reply .dotnet/CHANGELOG.md:3]: They're there; is GitHub pulling one of those "click here to load more files" things?

File: .dotnet/src/Custom/Chat/ChatCompletionOptions.cs (hunk 2 above)
Comment (trrwilson) [reply .dotnet/src/Custom/Chat/ChatCompletionOptions.cs:164]: Comments added. Metadata is customized to remove the public setter it otherwise has emitted for it.

File: .dotnet/src/Custom/Chat/ChatCompletionOptions.cs (hunk 3 above)
Comment (trrwilson) [reply .dotnet/src/Custom/Chat/ChatCompletionOptions.cs:167]: `AllowParallelToolCalls` is what I was attempting to pattern this off of; "Allow" just didn't feel like the right word since it's an explicitly directive "service, thou shalt store stuff" and not a "you can use parallel tool calls *if you want to*, model."

`StoreOutput` doesn't have a clear yes/no word in it for the boolean-ness; `Store` is confusable between being a noun and a verb on its own, and `Output` likewise has a lot of overloaded meanings. If were asked "what do you think a symbol named `StoreOutput` is?", my first guesses would be (a) a method; (b) a property providing an output from a store; and only then (c) a flag controlling whether output should be stored.

File: .dotnet/src/Custom/Chat/ChatMessageContent.cs (hunk 5 above)
Comment (trrwilson) [reply .dotnet/src/Custom/Chat/ChatMessageContent.cs:12]: It's not 100% clear to me what introduced the compilation break (this type was not otherwise changed, I don't believe), but without the change I reliably get this:
```
D:\s\joseharriaga-openai-in-typespec\.dotnet\src\Custom\Chat\ChatMessageContent.cs(12,11): error CS0121: The call is ambiguous between the following methods or properties: 'ChatMessageContent.ChatMessageContent(string)' and 'ChatMessageContent.ChatMessageContent(params ChatMessageContentPart[])' [D:\s\joseharriaga-openai-in-typespec\.dotnet\src\OpenAI.csproj::TargetFramework=netstandard2.0]
```

File: .dotnet/api/OpenAI.netstandard2.0.cs (hunk 7 above)
Comment (trrwilson) [reply .dotnet/api/OpenAI.netstandard2.0.cs:43]: Yes. Using export-api was very useful for all the new realtime types but it was very challenging with so much of this extra stuff present; I can pull that out into its own PR if it's too distracting -- just helped development a ton and figured it'd be worthwhile to keep.

File: .dotnet/src/Custom/Chat/ChatFunctionCall.cs (hunk 4 above)
Comment (trrwilson) [reply .dotnet/src/Custom/Chat/ChatFunctionCall.cs:45]: I can do that. This might be the most subtle customization rule yet, but it works and the old order *was* better:

```yaml
  - remark: preserve earlier, more intuitive field order
    patterns:
      - ChatCompletionRequestAssistantMessage.function_call
      - ChatCompletionResponseMessage.function_call
    type: >-
      {
        name: string;
        arguments: string;
      }
    assert_existing_type: >-
      {
        arguments: string;
        name: string;
      }
```

File: .typespec/streaming/models.tsp [hunk 10]
```diff
@@ -44,6 +44,9 @@ union AssistantStreamEvent {
 
 union ThreadStreamEvent {
   {
+    /** Whether to enable input audio transcription. */
+    enabled?: boolean,
```
Comment (joseharriaga) [inline .typespec/streaming/models.tsp:48]: Is this new property reflected anywhere in the library?

File: .dotnet/api/OpenAI.netstandard2.0.cs (hunk 7 above)
Comment (joseharriaga) [reply .dotnet/api/OpenAI.netstandard2.0.cs:43]: Got it. I was confused because I was using CodeFlow and for some reason it didn't pick up the changes to the Export-API.ps1 script.

I do think it would help to do it in a separate PR. 😊

File: .dotnet/src/Custom/Moderations/ModerationResult.cs [hunk 11]
```diff
@@ -8,19 +8,22 @@ namespace OpenAI.Moderations;
 [CodeGenModel("CreateModerationResponseResult")]
 [CodeGenSuppress("Categories")]
 [CodeGenSuppress("CategoryScores")]
-[CodeGenSuppress(nameof(ModerationResult), typeof(bool), typeof(InternalModerationCategories), typeof(InternalModerationCategoryScores))]
-[CodeGenSuppress(nameof(ModerationResult), typeof(bool), typeof(InternalModerationCategories), typeof(InternalModerationCategoryScores), typeof(IDictionary<string, BinaryData>))]
+[CodeGenSuppress("CategoryAppliedInputTypes")]
```
Comment (joseharriaga) [inline .dotnet/src/Custom/Moderations/ModerationResult.cs:11]: We need to add new methods to `ModerationClient` that can take the multimodal inputs as parameters. Should we do it here or in a separate PR?

File: .dotnet/src/Custom/Moderations/ModerationCategory.cs [hunk 12]
```diff
@@ -1,13 +1,17 @@
...
+    internal ModerationCategory(bool flagged, float score, ModerationFlaggedContentModalities flaggedContentModalities)
     {
         Flagged = flagged;
         Score = score;
+        FlaggedContentModalities = flaggedContentModalities;
     }
 
     public bool Flagged { get; }
     public float Score { get; }
+    public ModerationFlaggedContentModalities FlaggedContentModalities { get; }
```
Comment (joseharriaga) [inline .dotnet/src/Custom/Moderations/ModerationCategory.cs:16]: Two things:
1. In the context of the Moderations API, the term used to refer to the input is "input" instead of "content".
2. In the future, when we introduce a class to represent these multimodal inputs, I expect they will probably look very similar to `ChatMessageContentPart` and use the squished/spread union pattern, and similarly I also expect that there will be a `Kind` property.

With these two things in mind, I wonder if a better name would be: `FlaggedInputKinds`.

File: .dotnet/src/Custom/Moderations/ModerationFlaggedContentModalities.cs [hunk 13]
```diff
@@ -0,0 +1,13 @@
+using System;
+
+namespace OpenAI.Moderations;
+
+[Flags]
+public enum ModerationFlaggedContentModalities : int
```
Comment (joseharriaga) [inline .dotnet/src/Custom/Moderations/ModerationFlaggedContentModalities.cs:6]: When we add support for multimodal inputs, we should confirm that, contrary to the examples in OpenAI's documentation, these flags are only set when the `Flagged` property of `ModerationResult` is `true`. Otherwise, having the word "Flagged" as part of this enum's name would be misleading.

File: .dotnet/CHANGELOG.md [hunk 14]
```diff
@@ -1,5 +1,23 @@
...
+
+### Features added
+
+- `ChatCompletionOptions` now has an `EnableOutputStorage` property (reflecting [the REST `store` property](https://platform.openai.com/docs/api-reference/chat/create#chat-create-store)) as well as a `Metadata` property ([`metadata` in REST](https://platform.openai.com/docs/api-reference/chat/create#chat-create-metadata))
+- `ChatTokenUsage` (`Usage` on `ChatCompletion`) has been expanded:
+  - A new `InputTokenDetails` property (of new type `ChatInputTokenUsageDetails`) has been added, containing breakdowns of `AudioTokenCount` and `CachedTokenCount` for usage with supported models
+  - The existing `OutputTokenDetails` has been expanded to include `AudioTokenCount` for supported models
+- `ModerationResult` now includes `Illicit` and `IllictViolent` categories
+- `ModerationCategory` now includes a (bitmasked flag enum) `InputKinds` property, representing applied input modalities for the category
```
Comment (joseharriaga) [inline .dotnet/CHANGELOG.md:12]: `InputKinds` has a different name now.

File: .dotnet/src/Generated/Models/InternalCreateModerationResponseResultCategoryAppliedInputTypesSelfHarm1.cs [hunk 15]
```diff
@@ -0,0 +1,36 @@
+// <auto-generated/>
+
+#nullable disable
+
+using System;
+using System.ComponentModel;
+
+namespace OpenAI.Moderations
+{
+    internal readonly partial struct InternalCreateModerationResponseResultCategoryAppliedInputTypesSelfHarm1 : IEquatable<InternalCreateModerationResponseResultCategoryAppliedInputTypesSelfHarm1>
```
Comment (joseharriaga) [inline .dotnet/src/Generated/Models/InternalCreateModerationResponseResultCategoryAppliedInputTypesSelfHarm1.cs:10]: Curious: Do you know why this one is suffixed with a "1" while the others are not?

File: .dotnet/src/Generated/Models/InternalCreateModerationResponseResultCategoryAppliedInputTypes.cs [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+        public IReadOnlyList<string> HarassmentThreatening { get; }
+        public IReadOnlyList<string> Illicit { get; }
+        public IReadOnlyList<string> IllicitViolent { get; }
+        public IReadOnlyList<InternalCreateModerationResponseResultCategoryAppliedInputTypesSelfHarm1> SelfHarm { get; }
+        public IReadOnlyList<InternalCreateModerationResponseResultCategoryAppliedInputTypesSelfHarmIntent> SelfHarmIntent { get; }
+        public IReadOnlyList<InternalCreateModerationResponseResultCategoryAppliedInputTypesSelfHarmInstruction> SelfHarmInstructions { get; }
+        public IReadOnlyList<InternalCreateModerationResponseResultCategoryAppliedInputTypesSexual> Sexual { get; }
+        public IReadOnlyList<string> SexualMinors { get; }
+        public IReadOnlyList<InternalCreateModerationResponseResultCategoryAppliedInputTypesViolence> Violence { get; }
+        public IReadOnlyList<InternalCreateModerationResponseResultCategoryAppliedInputTypesViolenceGraphic> ViolenceGraphic { get; }
```
Comment (joseharriaga) [inline .dotnet/src/Generated/Models/InternalCreateModerationResponseResultCategoryAppliedInputTypes.cs:79]: nit: I was expecting all of these to follow the same pattern. Specifically, I would have guessed that all of them would be IReadOnlyList properties of some enum. The enum would have one ("text") or two ("text" or "image") possible value.

Could this be suggesting that there is something wrong with the TypeSpec in the case where there is only one possible value?

Comment (trrwilson) [reply .dotnet/src/Generated/Models/InternalCreateModerationResponseResultCategoryAppliedInputTypes.cs:79]: It looks that way, yes.

Comparing `Illict` (adopts `<string>`) and `SelfHarm (adopts `<enum>`):

```javascript
      /** The applied input type(s) for the category 'illicit'. */
      illicit: "text"[];
      /** The applied input type(s) for the category 'self-harm'. */
      `self-harm`: ("text" | "image")[];
```

Things that massage this:

- Changing `illicit` to `("text" | "text")[]` achieves the intended generation outcome of getting a type, but (a) it looks very weird; and (b) it emits non-compiling code (an enum with the value `Text` appearing twice)
- Changing it to `(string | "text")[]` makes it behave exactly as intended in .NET, but the origin spec really doesn't make this out to be a true extensible enum (and we don't treat it as such)

Comment (trrwilson) [reply .dotnet/src/Generated/Models/InternalCreateModerationResponseResultCategoryAppliedInputTypes.cs:79]: Looking at this... sigh... time to break out `options.yaml` again. We can make this clean (and better than the source) if we customize and formalize the commonality of the input types.

Comment (trrwilson) [reply .dotnet/src/Generated/Models/InternalCreateModerationResponseResultCategoryAppliedInputTypes.cs:79]: This still ends up emitting suboptimally without customization. I'm going to keep the customization because it's very nice for the spec, but there's likely some feature request fodder lurking here to make the emitter cleaner.

With `union ModerationAppliedInputType { string, text: "text", image: "image" }`:

- `foo: ModerationAppliedInputType.text[]` emits as an `IReadOnlyList<string>`
- `foo: (ModerationAppliedInputType.text | ModerationAppliedInputType.image)[]` emits as `IReadOnlyList<BinaryData>`
- `foo: ModerationAppliedInputType[]` emits as `IReadOnlyList<ModerationAppliedInputType>`

The last one is the nicest, but requires losing constraint information in the spec -- which we really don't want.

File: .dotnet/src/Generated/Models/InternalCreateModerationResponseResultCategoryAppliedInputTypesSelfHarm1.cs (hunk 15 above)
Comment (trrwilson) [reply .dotnet/src/Generated/Models/InternalCreateModerationResponseResultCategoryAppliedInputTypesSelfHarm1.cs]: Not entirely sure, but I think it has to do with the array item emplacement making things confused. This was mitigated in this specific instance with the spec update to extract/customize, though.

File: .dotnet/src/Custom/Moderations/ModerationCategory.cs (hunk 12 above)
Comment (trrwilson) [reply .dotnet/src/Custom/Moderations/ModerationCategory.cs:16]: Sure; per the [earlier comment](https://github.com/joseharriaga/openai-in-typespec/pull/301#discussion_r1807132492), I'm happy with that -- this is going to end up inconsistent with "modalities" representations in other endpoints, but I concur that there's risk associated with trying to fix the consistency -- risk that's really not high-reward.

File: .dotnet/src/Custom/Moderations/ModerationFlaggedContentModalities.cs (hunk 13 above)
Comment (trrwilson) [reply .dotnet/src/Custom/Moderations/ModerationFlaggedContentModalities.cs:7]: Given the spec and your earlier suggestion, I'll just back this up to `ModerationApplicableInputKinds`.

File: .dotnet/src/Custom/Moderations/ModerationResult.cs (hunk 11 above)
Comment (trrwilson) [reply .dotnet/src/Custom/Moderations/ModerationResult.cs:11]: Let's do it in a separate one. I added an item to the internal repository board.

File: .typespec/streaming/models.tsp (hunk 10 above)
Comment (trrwilson) [reply .typespec/streaming/models.tsp:48]: Not yet -- haven't investigated the Assistants changes and it looks suspicious to me. Filed a work item to track but I don't think we need to update the public surface yet.

File: .dotnet/api/OpenAI.netstandard2.0.cs (hunk 7 above)
Comment (trrwilson) [reply .dotnet/api/OpenAI.netstandard2.0.cs]: Moved it to
https://github.com/joseharriaga/openai-in-typespec/pull/311

File: .dotnet/src/Custom/Chat/ChatMessageContent.cs (hunk 5 above)
Comment (trrwilson) [reply .dotnet/src/Custom/Chat/ChatMessageContent.cs:12]: Yes, I was surprised... I'm curious about what exactly changed, but it didn't impact the public surface and there are too many fish being fried!

File: .dotnet/src/Custom/Chat/OpenAIChatModelFactory.cs [hunk 17]
```diff
@@ -169,8 +183,8 @@ public static StreamingChatCompletionUpdate StreamingChatCompletionUpdate(
     public static StreamingChatFunctionCallUpdate StreamingChatFunctionCallUpdate(string functionName = null, BinaryData functionArgumentsUpdate = null)
     {
         return new StreamingChatFunctionCallUpdate(
-            functionName,
-            functionArgumentsUpdate,
+            functionArgumentsUpdate: functionArgumentsUpdate,
+            functionName: functionName,
             serializedAdditionalRawData: null);
```
[the remark this answers is not in the document]
Comment (trrwilson) [reply .dotnet/src/Custom/Chat/OpenAIChatModelFactory.cs:188]: For sure -- added a rule and it's in.

**Review by joseharriaga — APPROVED** [review]
